packages/cli/src/util/build/import-builders.ts

```diff
@@ -0,0 +1,197 @@
+          buildersToAdd.add(spec);
+          continue;
+        }
+
+        if (
+          parsed.type === 'range' &&
+          !satisfies(builderPkg.version, parsed.rawSpec)
+        ) {
+          // An explicit Builder range was specified but it is not
+          // compatible with the version that is currently installed
+          output.debug(
+            `Installed version "${name}@${builderPkg.version}" is not compatible with "${parsed.rawSpec}"`
+          );
+          buildersToAdd.add(spec);
+          continue;
+        }
+
+        // TODO: handle `parsed.type === 'tag'` ("latest" vs. anything else?)
+
+        const path = join(dirname(pkgPath), builderPkg.main || 'index.js');
+        const builder = require(path);
+
+        results.set(spec, {
+          builder,
+          pkg: builderPkg,
+          path,
+          pkgPath,
+        });
+        output.debug(`Imported Builder "${name}" from "${dirname(pkgPath)}"`);
+      } catch (err: any) {
+        if (err.code === 'MODULE_NOT_FOUND') {
+          output.debug(`Failed to import "${name}": ${err}`);
+          buildersToAdd.add(spec);
+
+          if (err.message.includes('@now/build-utils')) {
+            const upgradeMessage =
+              parsed.scope === '@now'
+                ? ` Please update from "@now" to "@vercel".`
+                : '';
+            throw new Error(
+              `The legacy "${name}" Builder is not supported.${upgradeMessage}`
+            );
+          }
+        } else {
+          throw err;
+        }
+      }
+    }
+
+    // Add any Builders that are not yet present into `.vercel/builders`
+    if (buildersToAdd.size > 0) {
+      try {
+        const emptyPkgJson = {
+          private: true,
+          license: 'UNLICENSED',
+        };
+        await outputJSON(buildersPkgPath, emptyPkgJson, {
+          flag: 'wx',
+        });
+      } catch (err: any) {
+        if (err.code !== 'EEXIST') throw err;
+      }
+
+      output.debug(
+        `Installing Builders: ${Array.from(buildersToAdd).join(', ')}`
+      );
+      await spawnAsync(
+        'yarn',
+        ['add', '@vercel/build-utils', ...buildersToAdd],
+        { cwd: buildersDir }
+      );
+
+      // Cross-reference any builderSpecs from the saved `package.json` file,
+      // in case they were installed from a URL
+      const buildersPkg = await readJSONFile<PackageJson>(buildersPkgPath);
+      if (buildersPkg instanceof CantParseJSONFile) throw buildersPkg;
+      if (!buildersPkg) {
+        throw new Error(`Failed to load "${buildersPkgPath}"`);
+      }
+      for (const spec of buildersToAdd) {
+        for (const [name, version] of Object.entries(
+          buildersPkg.dependencies || {}
+        )) {
+          if (version === spec) {
+            output.debug(`Resolved Builder spec "${spec}" to name "${name}"`);
+            resolvedSpecs.set(spec, name);
+          }
+        }
+      }
+    }
+  } while (results.size !== builderSpecs.size);
```

Comment:
I *think* I follow this looping logic, but it's not easy. I think this should be made clearer at some point. I'd be happy to pair on it.

Reply:
I think the confusing part is that the `do {}` is so big so its not obvious its a do-while loop

Reply:
Refactored here: https://github.com/vercel/vercel/pull/7842